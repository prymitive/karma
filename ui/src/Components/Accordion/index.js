import React from "react";

import Collapsible from "react-collapsible";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";

const Trigger = ({ text, isOpen }) => (
  <div className="d-flex flex-row justify-content-between">
    <div>{text}</div>
    <div>
      <FontAwesomeIcon
        icon={isOpen ? faChevronDown : faChevronUp}
        className="text-muted"
      />
    </div>
  </div>
);

const Accordion = ({ text, content, extraProps }) => (
  <Collapsible
    triggerTagName="div"
    transitionTime={50}
    className="card"
    openedClassName="card"
    triggerClassName="card-header cursor-pointer"
    triggerOpenedClassName="card-header cursor-pointer bg-light"
    contentOuterClassName="collapse show"
    contentInnerClassName="card-body my-2"
    {...extraProps}
    trigger={<Trigger text={text} isOpen={false} />}
    triggerWhenOpen={<Trigger text={text} isOpen={true} />}
  >
    {content}
  </Collapsible>
);

export { Accordion };
