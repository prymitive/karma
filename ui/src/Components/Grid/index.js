import React from "react";
import PropTypes from "prop-types";

import { useObserver } from "mobx-react-lite";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { AlertGrid } from "./AlertGrid";
import { FatalError } from "./FatalError";
import { UpstreamError } from "./UpstreamError";
import { UpgradeNeeded } from "./UpgradeNeeded";
import { ReloadNeeded } from "./ReloadNeeded";
import { EmptyGrid } from "./EmptyGrid";

const Grid = ({ alertStore, settingsStore, silenceFormStore }) => {
  return useObserver(() =>
    alertStore.info.upgradeNeeded ? (
      <UpgradeNeeded newVersion={alertStore.info.version} reloadAfter={3000} />
    ) : alertStore.info.reloadNeeded ? (
      <ReloadNeeded reloadAfter={4000} />
    ) : alertStore.status.error ? (
      <FatalError message={alertStore.status.error} />
    ) : alertStore.data.upstreams.counters &&
      alertStore.data.upstreams.counters.total === 1 &&
      alertStore.data.upstreams.counters.healthy === 0 &&
      alertStore.data.upstreams.instances[0] &&
      alertStore.data.upstreams.instances[0].error !== "" ? (
      <FatalError message={alertStore.data.upstreams.instances[0].error} />
    ) : alertStore.info.version !== "unknown" &&
      alertStore.info.totalAlerts === 0 ? (
      <EmptyGrid />
    ) : (
      <React.Fragment>
        {alertStore.data.upstreams.instances
          .filter((upstream) => upstream.error !== "")
          .map((upstream) => (
            <UpstreamError
              key={upstream.name}
              name={upstream.name}
              message={upstream.error}
            />
          ))}
        <AlertGrid
          alertStore={alertStore}
          settingsStore={settingsStore}
          silenceFormStore={silenceFormStore}
        />
      </React.Fragment>
    )
  );
};
Grid.propTypes = {
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  settingsStore: PropTypes.instanceOf(Settings).isRequired,
  silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
};

export { Grid };
