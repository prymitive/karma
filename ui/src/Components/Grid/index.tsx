import { FC } from "react";

import { observer } from "mobx-react-lite";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import AlertGrid from "./AlertGrid";
import { FatalError } from "./FatalError";
import { UpgradeNeeded } from "./UpgradeNeeded";
import { ReloadNeeded } from "./ReloadNeeded";
import { EmptyGrid } from "./EmptyGrid";

const Grid: FC<{
  alertStore: AlertStore;
  silenceFormStore: SilenceFormStore;
  settingsStore: Settings;
}> = ({ alertStore, settingsStore, silenceFormStore }) => {
  return alertStore.info.upgradeNeeded ? (
    <UpgradeNeeded newVersion={alertStore.info.version} reloadAfter={3000} />
  ) : alertStore.info.reloadNeeded ? (
    <ReloadNeeded reloadAfter={4000} />
  ) : alertStore.status.error ? (
    <FatalError message={alertStore.status.error} />
  ) : alertStore.data.upstreams.counters &&
    alertStore.data.upstreams.counters.total === 1 &&
    alertStore.data.upstreams.counters.healthy === 0 &&
    alertStore.data.upstreams.instances[0] &&
    alertStore.data.upstreams.instances[0].error !== "" ? (
    <FatalError message={alertStore.data.upstreams.instances[0].error} />
  ) : alertStore.info.version !== "unknown" &&
    alertStore.info.totalAlerts === 0 ? (
    <EmptyGrid />
  ) : (
    <AlertGrid
      alertStore={alertStore}
      settingsStore={settingsStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

export default observer(Grid);
