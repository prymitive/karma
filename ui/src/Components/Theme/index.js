import React, { useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

const ThemeContext = React.createContext();

const BodyTheme = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
