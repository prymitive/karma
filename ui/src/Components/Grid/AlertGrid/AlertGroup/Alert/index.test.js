import React from "react";
import { act } from "react-dom/test-utils";

import { mount } from "enzyme";

import { advanceTo, clear } from "jest-date-mock";

import toDiffableHtml from "diffable-html";

import {
  MockAlert,
  MockAnnotation,
  MockAlertGroup,
  MockSilence,
} from "__mocks__/Alerts";
import { MockThemeContext } from "__mocks__/Theme";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { BorderClassMap } from "Common/Colors";
import { ThemeContext } from "Components/Theme";
import { Alert } from ".";

let alertStore;
let silenceFormStore;

beforeEach(() => {
  advanceTo(new Date(Date.UTC(2018, 7, 15, 20, 40, 0)));
  alertStore = new AlertStore([]);
  silenceFormStore = new SilenceFormStore();
});

afterEach(() => {
  // reset Date() to current time
  clear();
  jest.restoreAllMocks();
});

const MockAfterUpdate = jest.fn();
const MockSetIsMenuOpen = jest.fn();

const MockedAlert = () => {
  return MockAlert(
    [
      MockAnnotation("help", "some long text", true, false),
      MockAnnotation("hidden", "some hidden text", false, false),
      MockAnnotation("link", "http://localhost", true, true),
    ],
    { job: "node_exporter", cluster: "dev" },
    "active"
  );
};

const MountedAlert = (alert, group, showAlertmanagers, showReceiver) => {
  return mount(
    <Alert
      alert={alert}
      group={group}
      showAlertmanagers={showAlertmanagers}
      showReceiver={showReceiver}
      afterUpdate={MockAfterUpdate}
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
      setIsMenuOpen={MockSetIsMenuOpen}
    />,
    {
      wrappingComponent: ThemeContext.Provider,
      wrappingComponentProps: { value: MockThemeContext },
    }
  );
};

describe("<Alert />", () => {
  it("matches snapshot with showAlertmanagers=false showReceiver=false", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("matches snapshot when inhibited", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("renders inhibition icon when inhibited", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    alert.alertmanager.push({
      name: "ha2",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/prometheus",
      silencedBy: [],
      inhibitedBy: ["123456"],
    });
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(tree.find(".fa-volume-mute")).toHaveLength(1);
  });

  it("inhibition icon passes only unique fingerprints", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(tree.find(".fa-volume-mute")).toHaveLength(1);
  });

  it("renders @cluster label with showAlertmanagers=true", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, true, false);
    const label = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@cluster");
    expect(label.text()).toBe("@cluster: default");
  });

  it("only renders one @cluster label per alertmanager cluster", () => {
    const alert = MockedAlert();
    alert.alertmanager.push({
      name: "ha1",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/prometheus",
      silencedBy: [],
      inhibitedBy: [],
    });
    alert.alertmanager.push({
      name: "ha2",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/prometheus",
      silencedBy: [],
      inhibitedBy: [],
    });
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, true, false);
    const labels = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@cluster");
    expect(labels).toHaveLength(2);
    expect(labels.at(0).text()).toBe("@cluster: default");
    expect(labels.at(1).text()).toBe("@cluster: HA");
  });

  it("renders @receiver label with showReceiver=true", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, true);
    const label = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@receiver");
    expect(label.text()).toBe("@receiver: by-name");
  });

  it("renders a silence if alert is silenced", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.silences = {
      default: {
        silence123456789: MockSilence(),
      },
    };
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(1);
    expect(silence.html()).toMatch(/Mocked Silence/);
  });

  it("renders a fallback silence if the silence is not found in alertStore", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.silences = {
      default: {
        "123": MockSilence(),
      },
    };
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false);
    const silence = tree.find("FallbackSilenceDesciption");
    expect(silence).toHaveLength(1);
    expect(silence.html()).not.toMatch(/Mocked Silence/);
  });

  it("renders a fallback silence if the cluster is not found in alertStore", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.silences = {
      foo: {
        "123": MockSilence(),
      },
    };
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false);
    const silence = tree.find("FallbackSilenceDesciption");
    expect(silence).toHaveLength(1);
    expect(silence.html()).not.toMatch(/Mocked Silence/);
  });

  it("renders only one silence for HA cluster", () => {
    const alert = MockedAlert();
    alert.alertmanager = [
      {
        name: "am1",
        cluster: "ha",
        state: "suppressed",
        startsAt: "2018-08-14T17:36:40.017867056Z",
        source: "localhost/am1",
        silencedBy: ["silence123456789"],
        inhibitedBy: [],
      },
      {
        name: "am2",
        cluster: "ha",
        state: "suppressed",
        startsAt: "2018-08-14T17:36:40.017867056Z",
        source: "localhost/am2",
        silencedBy: ["silence123456789"],
        inhibitedBy: [],
      },
    ];
    alertStore.data.silences = {
      ha: {
        silence123456789: MockSilence(),
      },
    };
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(1);
    expect(silence.html()).toMatch(/Mocked Silence/);
  });

  it("doesn't render shared silences", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    const group = MockAlertGroup(
      {},
      [alert],
      [],
      {},
      { default: ["silence123456789"] }
    );
    const tree = MountedAlert(alert, group, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(0);
  });

  it("uses BorderClassMap.active when @state=active", () => {
    const alert = MockedAlert();
    alert.state = "active";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.active)
    ).toBe(true);
  });

  it("uses BorderClassMap.suppressed when @state=suppressed", () => {
    const alert = MockedAlert();
    alert.state = "suppressed";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.suppressed)
    ).toBe(true);
  });

  it("uses BorderClassMap.unprocessed when @state=unprocessed", () => {
    const alert = MockedAlert();
    alert.state = "unprocessed";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.unprocessed)
    ).toBe(true);
  });

  it("uses 'border-default' with unknown @state", () => {
    jest.spyOn(console, "error").mockImplementation(() => {});

    const alert = MockedAlert();
    alert.state = "foobar";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass("border-default")
    ).toBe(true);
  });

  it("alert timestamp is updated every minute", () => {
    jest.useFakeTimers();

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 36, 41)));

    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false);
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("just now");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 36, 42)));
    act(() => jest.advanceTimersByTime(31 * 1000));
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("a few seconds ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 37, 41)));
    act(() => jest.advanceTimersByTime(31 * 1000));
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("1 minute ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 18, 36, 41)));
    act(() => jest.advanceTimersByTime(31 * 1000));
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("1 hour ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 19, 36, 41)));
    act(() => jest.advanceTimersByTime(31 * 1000));
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("2 hours ago");

    advanceTo(new Date(Date.UTC(2018, 7, 16, 19, 36, 41)));
    act(() => jest.advanceTimersByTime(31 * 1000));
    expect(
      tree
        .find("span.components-label.badge.badge-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("2 days ago");
  });
});
