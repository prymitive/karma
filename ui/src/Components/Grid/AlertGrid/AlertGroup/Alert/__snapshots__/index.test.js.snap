// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> matches snapshot when inhibited 1`] = `
"
<li class=\\"components-grid-alertgrid-alertgroup-alert list-group-item bg-transparent pl-1 pr-0 py-0 my-1 rounded-0 border-left-1 border-right-0 border-top-0 border-bottom-0 border-danger\\">
  <div>
    <div style=\\"display: inline-block; max-width: 100%;\\"
         class=\\" tooltip-trigger\\"
    >
      <div class=\\"mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100\\">
        <span class=\\"cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               focusable=\\"false\\"
               data-prefix=\\"fas\\"
               data-icon=\\"search-minus\\"
               class=\\"svg-inline--fa fa-search-minus fa-w-16 mr-1\\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 512 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M304 192v32c0 6.6-5.4 12-12 12H124c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm201 284.7L476.7 505c-9.4 9.4-24.6 9.4-33.9 0L343 405.3c-4.5-4.5-7-10.6-7-17V372c-35.3 27.6-79.7 44-128 44C93.1 416 0 322.9 0 208S93.1 0 208 0s208 93.1 208 208c0 48.3-16.4 92.7-44 128h16.3c6.4 0 12.5 2.5 17 7l99.7 99.7c9.3 9.4 9.3 24.6 0 34zM344 208c0-75.2-60.8-136-136-136S72 132.8 72 208s60.8 136 136 136 136-60.8 136-136z\\"
            >
            </path>
          </svg>
          <span class=\\"text-muted\\">
            help:
          </span>
        </span>
        <span class=\\"Linkify\\">
          <span style=\\"opacity: 1;\\">
            some long text
          </span>
        </span>
      </div>
    </div>
    <div style=\\"display: inline-block; max-width: 100%;\\"
         class=\\" tooltip-trigger\\"
    >
      <div class=\\"mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100 cursor-pointer\\">
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"fas\\"
             data-icon=\\"search-plus\\"
             class=\\"svg-inline--fa fa-search-plus fa-w-16 mr-1\\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M304 192v32c0 6.6-5.4 12-12 12h-56v56c0 6.6-5.4 12-12 12h-32c-6.6 0-12-5.4-12-12v-56h-56c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h56v-56c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v56h56c6.6 0 12 5.4 12 12zm201 284.7L476.7 505c-9.4 9.4-24.6 9.4-33.9 0L343 405.3c-4.5-4.5-7-10.6-7-17V372c-35.3 27.6-79.7 44-128 44C93.1 416 0 322.9 0 208S93.1 0 208 0s208 93.1 208 208c0 48.3-16.4 92.7-44 128h16.3c6.4 0 12.5 2.5 17 7l99.7 99.7c9.3 9.4 9.3 24.6 0 34zM344 208c0-75.2-60.8-136-136-136S72 132.8 72 208s60.8 136 136 136 136-60.8 136-136z\\"
          >
          </path>
        </svg>
        hidden
      </div>
    </div>
  </div>
  <span>
    <span class=\\"components-label components-label-with-hover px-1 mr-1 badge badge-secondary cursor-pointer\\"
          data-toggle=\\"dropdown\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"caret-down\\"
           class=\\"svg-inline--fa fa-caret-down fa-w-10 pr-1\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 320 512\\"
           style=\\"width: 0.8rem;\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
        >
        </path>
      </svg>
      1 day ago
    </span>
  </span>
  <div style=\\"display: inline-block; max-width: 100%;\\"
       class=\\" tooltip-trigger\\"
  >
    <span class=\\"badge badge-light components-label components-label-with-hover cursor-pointer\\">
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"volume-mute\\"
           class=\\"svg-inline--fa fa-volume-mute fa-w-16 text-success\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 512 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M215.03 71.05L126.06 160H24c-13.26 0-24 10.74-24 24v144c0 13.25 10.74 24 24 24h102.06l88.97 88.95c15.03 15.03 40.97 4.47 40.97-16.97V88.02c0-21.46-25.96-31.98-40.97-16.97zM461.64 256l45.64-45.64c6.3-6.3 6.3-16.52 0-22.82l-22.82-22.82c-6.3-6.3-16.52-6.3-22.82 0L416 210.36l-45.64-45.64c-6.3-6.3-16.52-6.3-22.82 0l-22.82 22.82c-6.3 6.3-6.3 16.52 0 22.82L370.36 256l-45.63 45.63c-6.3 6.3-6.3 16.52 0 22.82l22.82 22.82c6.3 6.3 16.52 6.3 22.82 0L416 301.64l45.64 45.64c6.3 6.3 16.52 6.3 22.82 0l22.82-22.82c6.3-6.3 6.3-16.52 0-22.82L461.64 256z\\"
        >
        </path>
      </svg>
    </span>
  </div>
  <div style=\\"display: inline-block; max-width: 100%;\\"
       class=\\" tooltip-trigger\\"
  >
    <span class=\\"components-label badge badge-default components-label-dark components-label-with-hover\\">
      <span class=\\"components-label-name\\">
        job:
      </span>
      <span class=\\"components-label-value\\">
        node_exporter
      </span>
    </span>
  </div>
  <div style=\\"display: inline-block; max-width: 100%;\\"
       class=\\" tooltip-trigger\\"
  >
    <span class=\\"components-label badge badge-default components-label-dark components-label-with-hover\\">
      <span class=\\"components-label-name\\">
        cluster:
      </span>
      <span class=\\"components-label-value\\">
        dev
      </span>
    </span>
  </div>
  <a href=\\"http://localhost\\"
     target=\\"_blank\\"
     rel=\\"noopener noreferrer\\"
     class=\\"components-label components-label-with-hover badge components-grid-annotation-link\\"
  >
    <svg aria-hidden=\\"true\\"
         focusable=\\"false\\"
         data-prefix=\\"fas\\"
         data-icon=\\"external-link-alt\\"
         class=\\"svg-inline--fa fa-external-link-alt fa-w-16 \\"
         role=\\"img\\"
         xmlns=\\"http://www.w3.org/2000/svg\\"
         viewbox=\\"0 0 512 512\\"
    >
      <path fill=\\"currentColor\\"
            d=\\"M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z\\"
      >
      </path>
    </svg>
    link
  </a>
</li>
"
`;

exports[`<Alert /> matches snapshot with showAlertmanagers=false showReceiver=false 1`] = `
"
<li class=\\"components-grid-alertgrid-alertgroup-alert list-group-item bg-transparent pl-1 pr-0 py-0 my-1 rounded-0 border-left-1 border-right-0 border-top-0 border-bottom-0 border-danger\\">
  <div>
    <div style=\\"display: inline-block; max-width: 100%;\\"
         class=\\" tooltip-trigger\\"
    >
      <div class=\\"mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100\\">
        <span class=\\"cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               focusable=\\"false\\"
               data-prefix=\\"fas\\"
               data-icon=\\"search-minus\\"
               class=\\"svg-inline--fa fa-search-minus fa-w-16 mr-1\\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 512 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M304 192v32c0 6.6-5.4 12-12 12H124c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm201 284.7L476.7 505c-9.4 9.4-24.6 9.4-33.9 0L343 405.3c-4.5-4.5-7-10.6-7-17V372c-35.3 27.6-79.7 44-128 44C93.1 416 0 322.9 0 208S93.1 0 208 0s208 93.1 208 208c0 48.3-16.4 92.7-44 128h16.3c6.4 0 12.5 2.5 17 7l99.7 99.7c9.3 9.4 9.3 24.6 0 34zM344 208c0-75.2-60.8-136-136-136S72 132.8 72 208s60.8 136 136 136 136-60.8 136-136z\\"
            >
            </path>
          </svg>
          <span class=\\"text-muted\\">
            help:
          </span>
        </span>
        <span class=\\"Linkify\\">
          <span style=\\"opacity: 1;\\">
            some long text
          </span>
        </span>
      </div>
    </div>
    <div style=\\"display: inline-block; max-width: 100%;\\"
         class=\\" tooltip-trigger\\"
    >
      <div class=\\"mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100 cursor-pointer\\">
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"fas\\"
             data-icon=\\"search-plus\\"
             class=\\"svg-inline--fa fa-search-plus fa-w-16 mr-1\\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M304 192v32c0 6.6-5.4 12-12 12h-56v56c0 6.6-5.4 12-12 12h-32c-6.6 0-12-5.4-12-12v-56h-56c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h56v-56c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v56h56c6.6 0 12 5.4 12 12zm201 284.7L476.7 505c-9.4 9.4-24.6 9.4-33.9 0L343 405.3c-4.5-4.5-7-10.6-7-17V372c-35.3 27.6-79.7 44-128 44C93.1 416 0 322.9 0 208S93.1 0 208 0s208 93.1 208 208c0 48.3-16.4 92.7-44 128h16.3c6.4 0 12.5 2.5 17 7l99.7 99.7c9.3 9.4 9.3 24.6 0 34zM344 208c0-75.2-60.8-136-136-136S72 132.8 72 208s60.8 136 136 136 136-60.8 136-136z\\"
          >
          </path>
        </svg>
        hidden
      </div>
    </div>
  </div>
  <span>
    <span class=\\"components-label components-label-with-hover px-1 mr-1 badge badge-secondary cursor-pointer\\"
          data-toggle=\\"dropdown\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"caret-down\\"
           class=\\"svg-inline--fa fa-caret-down fa-w-10 pr-1\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 320 512\\"
           style=\\"width: 0.8rem;\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
        >
        </path>
      </svg>
      1 day ago
    </span>
  </span>
  <div style=\\"display: inline-block; max-width: 100%;\\"
       class=\\" tooltip-trigger\\"
  >
    <span class=\\"components-label badge badge-default components-label-dark components-label-with-hover\\">
      <span class=\\"components-label-name\\">
        job:
      </span>
      <span class=\\"components-label-value\\">
        node_exporter
      </span>
    </span>
  </div>
  <div style=\\"display: inline-block; max-width: 100%;\\"
       class=\\" tooltip-trigger\\"
  >
    <span class=\\"components-label badge badge-default components-label-dark components-label-with-hover\\">
      <span class=\\"components-label-name\\">
        cluster:
      </span>
      <span class=\\"components-label-value\\">
        dev
      </span>
    </span>
  </div>
  <a href=\\"http://localhost\\"
     target=\\"_blank\\"
     rel=\\"noopener noreferrer\\"
     class=\\"components-label components-label-with-hover badge components-grid-annotation-link\\"
  >
    <svg aria-hidden=\\"true\\"
         focusable=\\"false\\"
         data-prefix=\\"fas\\"
         data-icon=\\"external-link-alt\\"
         class=\\"svg-inline--fa fa-external-link-alt fa-w-16 \\"
         role=\\"img\\"
         xmlns=\\"http://www.w3.org/2000/svg\\"
         viewbox=\\"0 0 512 512\\"
    >
      <path fill=\\"currentColor\\"
            d=\\"M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z\\"
      >
      </path>
    </svg>
    link
  </a>
</li>
"
`;
